### PR TITLE
Use PyGithub to reuse auth headers to do requests

### PR DESCRIPTION
**Pull request by bwrsandman**
Clean-up custom url creation and use PyGithub objects for that.
Use PyGithub checker function to catch 400+ codes.

This fixes app authenticated sessions losing the authentication for requests when downloading diffs and posting reviews.
This fixes `~/.netrc` authentication hijacking the authentication through the requests library.

Re-use the hashable comment class to filter already posted comments

### CODE REVIEW

**Comment by bwrsandman** [comment]
@ZedThree this is probably the last set of large PRs for my needs

**Comment by bwrsandman** [comment]
Worth noting that you need specific permissions for an app installation
* Contents (Read): Download the diff
* Pull requests (Read and Write): Read and post Review comments

**Comment by bwrsandman** [comment]
I tried getting the pr diff with `'pull_request.diff_url` which is `'https://github.com/<ORG>/<REPO>/pull/<PR_NUMBER>.diff'` but the response is empty.
It could be due to the redirection done with githubusercontent.com or something else.

File: post/clang_tidy_review/clang_tidy_review/__init__.py [hunk 1]
```diff
@@ -906,14 +909,11 @@ def download_artifacts(pull: PullRequest, workflow_id: int):
         )
         return None, None
 
-    r = requests.get(artifact.archive_download_url, headers=pull.headers("json"))
-    if not r.ok:
-        print(
-            f"WARNING: Couldn't automatically download artifacts for workflow '{workflow_id}', response was: {r}: {r.reason}"
-        )
-        return None, None
+    _, data = pull.repo._requester.requestJsonAndCheck(
```
Comment (ZedThree) [inline post/clang_tidy_review/clang_tidy_review/__init__.py:912]: Might be nice to retain the more specific error message if this fails?

File: post/clang_tidy_review/clang_tidy_review/__init__.py [hunk 2]
```diff
@@ -1049,19 +1021,14 @@ def cull_comments(pull_request: PullRequest, review, max_comments):
...
-            )
-        )
+    unposted_comments = set(map(lambda c: HashableComment(**c), review["comments"]))
+    posted_comments = set(
+        map(lambda c: HashableComment(**c), pull_request.get_pr_comments())
+    )
+
+    review["comments"] = [
+        c.__dict__ for c in sorted(unposted_comments - posted_comments)
+    ]
```
Comment (ZedThree) [inline post/clang_tidy_review/clang_tidy_review/__init__.py:1031]: This is a lot clearer, thanks!

**Comment by ZedThree** [comment]
> I tried getting the pr diff with `pull_request.diff_url` which is `https://github.com/<ORG>/<REPO>/pull/<PR_NUMBER>.diff` but the response is empty. It could be due to the redirection done with githubusercontent.com or something else.

Is this something that still needs looking at, or is the PR ready as is?

**Comment by bwrsandman** [comment]
> Is this something that still needs looking at, or is the PR ready as is?
It's very minor. I wanted to reduce the custom API stuff but looks like it's needed here.

I am going to put the PR in draft just so I can test it but I will put it back to read today and you can merge/tag.

**Comment by bwrsandman** [comment]
@ZedThree is there anything left to do ?

**Review by ZedThree — APPROVED** [review]
Thanks for your work on this @bwrsandman !

**Pull request merged**